feat: Add logging to explain reasons for Comet not being able to run a query stage natively

andygrove opened the pull request:
## Which issue does this PR close?

<!--
We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax. For example `Closes #123` indicates that this PR will close issue #123.
-->

Closes https://github.com/apache/datafusion-comet/issues/385

## Rationale for this change

<!--
 Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.  
-->

As a user, I would like to understand why my query stages are not running natively.

## What changes are included in this PR?

<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

New config + logging

## How are these changes tested?

<!--
We typically require tests for all PRs in order to:
1. Prevent the code from being accidentally broken by subsequent changes
2. Serve as another way to document the expected behavior of the code

If tests are not included in your PR, please explain why (for example, are they covered by existing tests)?
-->

Tested manually:

```
scala> spark.read.parquet("/tmp/test").createOrReplaceTempView("t1")
24/05/07 10:55:16 INFO src/lib.rs: Comet native library initialized
24/05/07 10:55:16 WARN CometSparkSessionExtensions$CometExecRule: Comet cannot execute this plan natively because Execute CreateViewCommand is not supported

scala> spark.sql("select * from t1 where cast(cast(a as double) as decimal(10,2)) > 5").show
24/05/07 10:55:29 WARN CometSparkSessionExtensions$CometExecRule: Comet cannot execute this plan natively because:
	- Comet does not guarantee correct results for cast from DoubleType to DecimalType(10,2) with timezone Some(America/Denver) and evalMode LEGACY (No overflow check). To enable all incompatible casts, set spark.comet.cast.allowIncompatible=true
	- CollectLimit is not supported
+---+
|  a|
+---+
|  9|
|  8|
|  6|
|  7|
+---+
```

andygrove commented on common/src/main/scala/org/apache/comet/CometConf.scala:
```diff
@@ -276,6 +276,14 @@ object CometConf {
+        "Set this config to VERBOSE to see logging explaining the reason(s) why a query " +
+          "stage cannot be executed natively with Comet.")
+      .stringConf
+      .createWithDefault("VERBOSE")
```
I'm still thinking about what other options we may want here, or if this should just be a boolean config

andygrove commented:
@parthchandra How does this look?

parthchandra commented on common/src/main/scala/org/apache/comet/CometConf.scala:
```diff
@@ -276,6 +276,14 @@ object CometConf {
       .booleanConf
       .createWithDefault(false)
 
+  val COMET_EXPLAIN_FALLBACK_ENABLED: ConfigEntry[Boolean] =
```
Do we really need to make this configurable?

viirya replied:
I think a config should be good. Sometimes it might be verbose.

parthchandra commented on spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala:
```diff
@@ -734,6 +734,22 @@ class CometSparkSessionExtensions
       } else {
         var newPlan = transform(plan)
 
+        // if the plan cannot be run natively then explain why (when appropriate config is enabled)
```
Given that ExecRule is going to be invoked multiple times as planning proceeds, this may get logged multiple times for the same plan. Perhaps that is why we need to keep this configurable.

andygrove replied:
That is one reason, yes. The logging could get verbose.

andygrove replied:
I would assume that this feature is useful in development but less so in production since we have the integration with Spark's explain for that.

parthchandra reviewed (approved):
LGTM. Just some observations.

viirya commented on spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala:
```diff
@@ -734,6 +734,22 @@ class CometSparkSessionExtensions
         var newPlan = transform(plan)
 
+        // if the plan cannot be run natively then explain why (when appropriate config is enabled)
+        if (CometConf.COMET_EXPLAIN_FALLBACK_ENABLED.get() && !isCometNative(newPlan)) {
```
Comet could only trigger native execution for partial query plan. This check `isCometNative(newPlan)` only applies for the top operator.

andygrove commented:
I am moving this to draft for now while I test with TPC-H queries and make some improvements

andygrove commented:
@viirya @parthchandra This PR is ready for another review. I simplified the logic and removed the check to see if the top level operator is native or not. Instead, if we have recorded any info using `withInfo` then that info will be displayed when the config is enabled. There is an example in the documentation as part of this PR.

kazuyukitanimura commented on spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala:
```diff
@@ -734,6 +734,23 @@ class CometSparkSessionExtensions
+        // if the plan cannot be run fully natively then explain why (when appropriate
+        // config is enabled)
+        if (CometConf.COMET_EXPLAIN_FALLBACK_ENABLED.get()) {
+          new ExtendedExplainInfo().extensionInfo(newPlan) match {
```
Unrelated, but I feel `ExtendedExplainInfo` can be `object`

andygrove replied:
Yes, I had the same thought

andygrove replied:
It is possible that it has to be a class due to the way this integrates into Spark, but I am not sure.

parthchandra replied:
Just seeing this. I don't think Spark prevents us from this being an object. I'll take care of this. https://github.com/apache/datafusion-comet/issues/452

kazuyukitanimura commented on spark/src/main/scala/org/apache/comet/ExtendedExplainInfo.scala:
```diff
@@ -45,7 +45,7 @@ class ExtendedExplainInfo extends ExtendedExplainGenerator {
   }
 
-  private def extensionInfo(node: TreeNode[_]): Set[String] = {
+  def extensionInfo(node: TreeNode[_]): Set[String] = {
```
Perhaps `private[comet]` instead of public?

andygrove commented:
@viirya I have 2 approvals but need a committer approval to be able to merge.

viirya commented:
Looks good to me. Thanks @andygrove
